dhis-2/dhis-web-server/src/main/java/org/hisp/dhis/web/appbundler/AppBundler.java

```diff
@@ -0,0 +1,438 @@
+  private static final String BUNDLE_INFO_FILE = "apps-bundle.json";
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  private final String downloadDir;
+  private final String artifactId;
+  private final String appListFilePath;
+  private final String defaultBranch;
+  private final AppBundleInfo bundleInfo;
+  private final String buildDir;
+
+  /**
+   * Constructs an AppBundler with the specified parameters.
+   *
+   * @param downloadDir the directory where the app ZIP files will be stored
+   * @param artifactId the artifact ID for the bundle
+   * @param appListFilePath the path to the JSON file containing the list of apps to bundle
+   * @param defaultBranch the default branch to use if none is specified in the app URL
+   */
+  public AppBundler(
+      String downloadDir,
+      String buildDir,
+      String artifactId,
+      String appListFilePath,
+      String defaultBranch) {
+    this.downloadDir = downloadDir;
+    this.buildDir = buildDir;
+    this.artifactId = artifactId;
+    this.appListFilePath = appListFilePath;
+    this.defaultBranch = defaultBranch != null ? defaultBranch : DEFAULT_BRANCH;
+    this.bundleInfo = new AppBundleInfo();
+  }
+
+  /**
+   * Record to hold parsed information from a standard GitHub URL.
+   *
+   * @param owner The repository owner.
+   * @param repo The repository name.
+   * @param ref The branch, tag, or commit reference.
+   * @param originalUrl The original URL string from the input file.
+   * @param codeloadUrl The converted URL for downloading the ZIP archive.
+   */
+  private record AppGithubRepo(
+      String owner, String repo, String ref, String originalUrl, String codeloadUrl) {}
+
+  /**
+   * Executes the app bundling process.
+   *
+   * @throws IOException if there's an error reading the app list or downloading the apps
+   */
+  public void execute() throws IOException {
+    log.info("Starting app bundling process");
+
+    Path downloadDirPath = Paths.get(downloadDir);
+    Files.createDirectories(downloadDirPath);
+
+    Path artifactDirPath = downloadDirPath.resolve(artifactId);
+    Files.createDirectories(artifactDirPath);
+
+    Path checksumDirPath = artifactDirPath.resolve(CHECKSUM_DIR);
+    Files.createDirectories(checksumDirPath);
+
+    List<AppGithubRepo> appRepoInfos = parseAppListUrls(appListFilePath);
+
+    log.info("Found {} valid apps to bundle", appRepoInfos.size());
+
+    // Download each app in parallel
+    downloadApps(appRepoInfos, artifactDirPath, checksumDirPath);
+
+    // Copy downloaded apps to build directory
+    Path targetArtifactDir = copyAppsToBuildDir(downloadDirPath);
+
+    // Write bundle info file
+    writeBundleFile(targetArtifactDir);
+
+    log.info("App bundling process completed successfully");
+  }
+
+  private void downloadApps(
+      List<AppGithubRepo> appRepoInfos, Path artifactDirPath, Path checksumDirPath) {
+    ExecutorService executor = Executors.newFixedThreadPool(DOWNLOAD_POOL_SIZE);
```

Comment:
You could also consider using `parallelStream` to make the download parallel without adding much code level ceremony.